server/athenian/api/auth.py

```diff
@@ -0,0 +1,118 @@
+import logging
+import os
+import re
+import json
+import aiohttp.web
+import dateutil.parser
+from jose import jwt
+
+# Format error response and append status code.
+class AuthError(Exception):
+    def __init__(self, error, status_code):
+        self.error = error
+        self.status_code = status_code
+
+class User():
+    def __init__(self, nickname: str, name: str, picture: str, updated_at: str, **kwargs):
+        self.username = nickname
+        self.fullname = name
+        self.picture = picture
+        self.updated = dateutil.parser.parse(updated_at)
+
+AUTH0_DOMAIN = os.getenv("AUTH0_DOMAIN")
+AUTH0_AUDIENCE = os.getenv("AUTH0_AUDIENCE")
+
+def auth_is_configured():
+    return AUTH0_DOMAIN and AUTH0_AUDIENCE
+
+class Auth0:
+    def __init__(self, domain=AUTH0_DOMAIN, audience=AUTH0_AUDIENCE, whitelist=tuple()):
+        self._domain = domain
+        self._audience = audience
+        self._whitelist = whitelist
+        self._session = aiohttp.ClientSession()
```

Comment:
So IMHO we should do the following for a clean exit:

1. Add `async def close(self)` to `Auth0`
2. Save `Auth0` instance inside `AthenianApp`
3. Add `async def close(self)` to `AthenianApp` which `await`-s `Auth0.close()`
4. In `def main()`:

```python
try:
    app.run(...)
finally:
    asyncio.get_event_loop().run_until_complete(app.close())
```

Reply:
Done, getting new errors now:
```
RuntimeError: Event loop is closed
sys:1: RuntimeWarning: coroutine 'AthenianApp.close' was never awaited
ERROR:7224:asyncio:Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7fa60e01afd0>
ERROR:7224:asyncio:Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x7fa5ee86ba70>, 403408.321548599)]']
connector: <aiohttp.connector.TCPConnector object at 0x7fa5f4ea2b70>
```

Reply:
OK let's leave it like this. I will fix it after merging, this requires some thought. I know why it happened but the fix requires some experimentation.